File: cpace-protoss-comparison/dalek/src/protoss_protocol.rs
#![forbid(unsafe_code)]

use std::fmt;
use curve25519_dalek::{
    ristretto::RistrettoPoint,
    scalar::Scalar,
};
use getrandom::getrandom;
use sha2::{Digest, Sha512};

use crate::ec_operations::{
    hash_to_point, concatenate_vectors,
    SCALAR_LEN,
};

pub const SESSION_KEY_BYTES: usize = 32; // Size of session key
pub const SESSION_ID_BYTES: usize = 16;

#[derive(Debug)]
pub enum Error {
    Random(getrandom::Error),
    InvalidPoint,
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Error::Random(e) => write!(f, "Random number generation failed: {:?}", e),
            Error::InvalidPoint => write!(f, "Invalid Ristretto point"),
        }
    }
}

impl From<getrandom::Error> for Error {
    fn from(e: getrandom::Error) -> Self {
        Error::Random(e)
    }
}

// Struct to represent ProtossState
#[derive(Clone)]
pub struct ProtossState {
    pub x: Scalar,
    pub i: RistrettoPoint,
    pub p_i: [u8; SESSION_ID_BYTES],
    pub p_j: [u8; SESSION_ID_BYTES],
    pub v: RistrettoPoint,
}

// Structs for return types
#[derive(Clone)]
pub struct ReturnTypeInit {
    pub i: RistrettoPoint,
    pub state: ProtossState,
}

#[derive(Clone)]
pub struct ReturnTypeRspDer {
    pub r: RistrettoPoint,
    pub k: [u8; SESSION_KEY_BYTES],
}

// Initialize protocol
pub fn init(
    password: &str,
    p_i: &[u8; SESSION_ID_BYTES],
    p_j: &mut [u8; SESSION_ID_BYTES],
) -> Result<ReturnTypeInit, Error> {
    // Choose random x on elliptic curve
    let mut x_bytes = [0u8; SCALAR_LEN];
    getrandom(&mut x_bytes)?;
    let x = Scalar::from_bytes_mod_order(x_bytes);

    // Calculate X = g^x
    let _x_point = RistrettoPoint::mul_base(&x);

    // Calculate V = Hash(pwd), where V must be a point on the curve
    let v = hash_to_point(password);

    // Calculate I = X*V ~> X + V in elliptic curve calculations
    let i = _x_point + v;

    // Create state
    let state = ProtossState {
        x,
        i,
        p_i: *p_i,
        p_j: *p_j,
        v,
    };

    Ok(ReturnTypeInit { i, state })
}

// Responder derivation
pub fn rsp_der(
    password: &str,
    p_i: &[u8; SESSION_ID_BYTES],
    p_j: &mut [u8; SESSION_ID_BYTES],
    i: RistrettoPoint,
) -> Result<ReturnTypeRspDer, Error> {
    // Choose random y on elliptic curve
    let mut y_bytes = [0u8; SCALAR_LEN];
    getrandom(&mut y_bytes)?;

    // Maps to random element on Z_p
    let y = Scalar::from_bytes_mod_order(y_bytes);

    // Calculate Y = g^y
    let _y_point = RistrettoPoint::mul_base(&y);

    // Calculate V = Hash(pwd)
    let v = hash_to_point(password);

    // Calculate R = Y*V ~> Y + V in elliptic curve calculations
    let r = _y_point + v;

    // Calculate X' = I/V ~> I - V, because I and V are elliptic curve points
    let x_prime = i - v;

    // Calculate Z = (X')^y ~> y*X' in elliptic curve calculations
    let z = y*x_prime;

    // Calculate K = H'(Z, I, R, P_i, P_j, V)
    let concat = concatenate_vectors(&[
        z.compress().as_bytes(),
        i.compress().as_bytes(),
        r.compress().as_bytes(),
        p_i.as_ref(),
        p_j.as_ref(),
        v.compress().as_bytes(),
    ]);
    let mut hasher = Sha512::new();
    hasher.update(&concat);
    let hash = hasher.finalize();
    let mut k = [0u8; SESSION_KEY_BYTES];
    k.copy_from_slice(&hash[..SESSION_KEY_BYTES]);

    Ok(ReturnTypeRspDer { r, k })
}

// Derivation
pub fn der(
    _password: &str, 
    protoss_state: ProtossState,
    r: RistrettoPoint,
) -> Result<[u8; SESSION_KEY_BYTES], Error> {
    let ProtossState { x, i, p_i, p_j, v } = protoss_state;

    // Calculate Y' = R/V ~> R - V because R and V are elliptic curve points
    let y_prime = r - v;

    // Calculated Z = (Y')^x ~> x*Y' in elliptic curve calcuations
    let z = x*y_prime;

    // Calculate K = H'(Z, I, R, P_i, P_j, V)
    let concat = concatenate_vectors(&[
        z.compress().as_bytes(),
        i.compress().as_bytes(),
        r.compress().as_bytes(),
        p_i.as_ref(),
        p_j.as_ref(),
        v.compress().as_bytes(),
    ]);
    let mut hasher = Sha512::new();
    hasher.update(&concat);
    let hash = hasher.finalize();
    let mut k = [0u8; SESSION_KEY_BYTES];
    k.copy_from_slice(&hash[..SESSION_KEY_BYTES]);

    Ok(k)
}

#[cfg(test)]
mod tests {
    use super::*;
    use hex::encode;

    #[test]
    fn test_protoss_protocol() {
        // Initialize inputs
        let password = "my_secure_password";
        let p_i = [1u8; SESSION_ID_BYTES];
        let mut p_j = [2u8; SESSION_ID_BYTES];

        // Step 1: Init
        let init_result = init(password, &p_i, &mut p_j).unwrap();
        let i = init_result.i;

        // Step 2: Responder derivation
        let rsp_der_result = rsp_der(password, &p_i, &mut p_j, i).unwrap();
        let r = rsp_der_result.r;
        let k1 = rsp_der_result.k;

        // Step 3: Derivation
        let k2 = der(password, init_result.state, r).unwrap();

        // Verify keys match
        assert_eq!(k1, k2, "Session keys do not match");
        println!("Session key: {}", encode(k1));
    }
} 

File: libsodium-c/src/protoss_protocol.c

#include "protoss_protocol.h"
#include <string.h>

// Hash password -> 64-byte hash -> map to Ristretto point
int hash_to_point(unsigned char out[PROTOSS_POINT_LEN],
                  const char *password, size_t password_len)
{
    unsigned char hash[PROTOSS_HASH_INPUT_LEN];
    if (crypto_generichash(hash, sizeof(hash),
                           (const unsigned char *)password, password_len,
                           NULL, 0) != 0)
        return -1;

    if (crypto_core_ristretto255_from_hash(out, hash) != 0)
        return -1;

    return 0;
}

// Derive session key K = H'(Z, I, R, P_i, P_j, V) using streaming hash
static int derive_session_key(unsigned char K[PROTOSS_SESSION_KEY_LEN],
                              const unsigned char Z[PROTOSS_POINT_LEN],
                              const unsigned char I[PROTOSS_POINT_LEN],
                              const unsigned char R[PROTOSS_POINT_LEN],
                              const unsigned char *P_i, size_t P_i_len,
                              const unsigned char *P_j, size_t P_j_len,
                              const unsigned char V[PROTOSS_POINT_LEN])
{
    crypto_generichash_state h_state;
    if (crypto_generichash_init(&h_state, NULL, 0, PROTOSS_SESSION_KEY_LEN) != 0)
        return -1;
    crypto_generichash_update(&h_state, Z, PROTOSS_POINT_LEN);
    crypto_generichash_update(&h_state, I, PROTOSS_POINT_LEN);
    crypto_generichash_update(&h_state, R, PROTOSS_POINT_LEN);
    crypto_generichash_update(&h_state, P_i, P_i_len);
    crypto_generichash_update(&h_state, P_j, P_j_len);
    crypto_generichash_update(&h_state, V, PROTOSS_POINT_LEN);
    if (crypto_generichash_final(&h_state, K, PROTOSS_SESSION_KEY_LEN) != 0)
        return -1;
    return 0;
}

int Init(ReturnTypeInit *result,
         const char *password, size_t password_len,
         const unsigned char *P_i, size_t P_i_len,
         const unsigned char *P_j, size_t P_j_len)
{
    unsigned char x[PROTOSS_SCALAR_LEN];
    unsigned char X[PROTOSS_POINT_LEN];
    unsigned char V[PROTOSS_POINT_LEN];
    unsigned char I[PROTOSS_POINT_LEN];

    // choose random x in Z_p
    crypto_core_ristretto255_scalar_random(x);

    // calculate X = g^x
    if (crypto_scalarmult_ristretto255_base(X, x) != 0)
        return -1;

    // Calculate V = Hash(pwd)
    if (hash_to_point(V, password, password_len) != 0)
        return -1;

    // Calculate I = X*V ~> X + V in elliptic curves
    if (crypto_core_ristretto255_add(I, X, V) != 0)
        return -1;

    // Populate result
    memcpy(result->I, I, PROTOSS_POINT_LEN);
    memcpy(result->state.x, x, PROTOSS_SCALAR_LEN);
    memcpy(result->state.I, I, PROTOSS_POINT_LEN);
    memcpy(result->state.V, V, PROTOSS_POINT_LEN);
    memset(result->state.P_i, 0, PROTOSS_MAX_ID_LEN);
    memcpy(result->state.P_i, P_i, P_i_len);
    result->state.P_i_len = P_i_len;
    memset(result->state.P_j, 0, PROTOSS_MAX_ID_LEN);
    memcpy(result->state.P_j, P_j, P_j_len);
    result->state.P_j_len = P_j_len;

    return 0;
}

int RspDer(ReturnTypeRspDer *result,
           const char *password, size_t password_len,
           const unsigned char *P_i, size_t P_i_len,
           const unsigned char *P_j, size_t P_j_len,
           const unsigned char I[PROTOSS_POINT_LEN])
{
    unsigned char y[PROTOSS_SCALAR_LEN];
    unsigned char Y[PROTOSS_POINT_LEN];
    unsigned char V[PROTOSS_POINT_LEN];
    unsigned char R[PROTOSS_POINT_LEN];
    unsigned char X_prime[PROTOSS_POINT_LEN];
    unsigned char Z[PROTOSS_POINT_LEN];

    // Choose random y in Z_p
    crypto_core_ristretto255_scalar_random(y);

    // Calculate Y = g^y
    if (crypto_scalarmult_ristretto255_base(Y, y) != 0)
        return -1;

    // Calculate V = Hash(pwd)
    if (hash_to_point(V, password, password_len) != 0)
        return -1;

    // Calculate R = Y*V  ~> Y + V on the elliptic curve
    if (crypto_core_ristretto255_add(R, Y, V) != 0)
        return -1;

    // Calculates X' = I/V ~> I - V, because I and V are elliptic curve points
    if (crypto_core_ristretto255_sub(X_prime, I, V) != 0)
        return -1;

    // Calculates Z = (X')^y ~> y*X' in elliptic curve calculations
    if (crypto_scalarmult_ristretto255(Z, y, X_prime) != 0)
        return -1;

    // Calculates K = H'(Z, I, R, P_i, P_j, V)
    if (derive_session_key(result->K, Z, I, R, P_i, P_i_len, P_j, P_j_len, V) != 0)
        return -1;

    memcpy(result->R, R, PROTOSS_POINT_LEN);
    return 0;
}

int Der(unsigned char K[PROTOSS_SESSION_KEY_LEN],
        const ProtossState *state,
        const unsigned char R[PROTOSS_POINT_LEN])
{
    unsigned char Y_prime[PROTOSS_POINT_LEN];
    unsigned char Z[PROTOSS_POINT_LEN];

    // Calculate Y' = R/V ~> R - V because R and V are elliptic curve points
    if (crypto_core_ristretto255_sub(Y_prime, R, state->V) != 0)
        return -1;

    // Calculates Z = (Y')^x ~> x*Y' in elliptic curve calculations
    if (crypto_scalarmult_ristretto255(Z, state->x, Y_prime) != 0)
        return -1;

    // Calculates K = H'(Z, I, R, P_i, P_j, V)
    if (derive_session_key(K, Z, state->I, R,
                           state->P_i, state->P_i_len,
                           state->P_j, state->P_j_len,
                           state->V) != 0)
        return -1;

    return 0;
}
